### PR TITLE
[Workflow] Method buildTransitionBlockerList returns TransitionBlockerList of expected transition

### PR DESCRIPTION
**Pull request by Tetragramat**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28432, #28429
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by Tetragramat** [comment]
Why it was not merged into 4.1 already? Is there some problem with this PR? 
It is bugfix of buildTransitionBlockerList method which was not fully covered by unit tests. Thanks to that it contained sneaky bug.

**Comment by Tetragramat** [comment]
ping @lyrixx

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you for this PR. Great that you added tests!

It is indeed a bug you've found. However, I'm not happy with the solution. It is correct, because transition names have to be unique for a place.

However I think a simpler solution would be: 

```php
if ($transitionBlockerList->isEmpty()) {
    // We found a transition that we are allowed to do. Ie, we are not blocked.
    return $transitionBlockerList;
}
```

It works because the `$marking->has($place)` for all "froms" is already done in `buildTransitionBlockerListForTransition`.

If you make this update, then Im happy to approve this change.

**Comment by Tetragramat** [comment]
I've found out that in case of state_machine transition names are not unique. At least that is how framework bundle generates transitons from configuration. see https://github.com/symfony/symfony/blob/v4.1.4/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L512

Your solution will not work correctly. Yes, you will get empty blocker list when expected to be empty, but in case of all blocked same name transitions, you will get wrong blocker list with wrong reasons why transition was blocked. Maybe I should add more test for this case.

**Comment by Tetragramat** [comment]
I tested only half of the issue. Now it should be fully covered.

**Comment by Tetragramat** [comment]
I have no idea where these failing tests came from, but it have nothing to do with my changes.

**Comment by Nyholm** [comment]
Hm. Transition names are always unique from a place. Check the Validators. I need a whiteboard to figure this out properly. =)

Since it is basically just you, (maybe I) and @lyrixx that understand this fix and the test after we look at it for 20 minutes, Could you add some more comments on your code change? Also a comment on the test to describe what they are testing?
I fear we and other developers will get an headache in the future if this is not properly documented.

**Comment by Tetragramat** [comment]
What validators I'm supposed to check? I see no workflow definition validators.

I tried as hard as I could to make descriptions. I'm no good at explaining and english either.

**Comment by xabbuh** [comment]
I agree with @Nyholm that having the `break` here is not really easy to understand. I suggest to solve it by a change like this: https://github.com/Tetragramat/symfony/compare/4.1...xabbuh:pr-28493

**Comment by Nyholm** [comment]
Thank you for adding the comments and tests. What do you (@Tetragramat) think about @xabbuh's solution to the same problem?

It makes the code a bit easier to follow, right?

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php [hunk 1]
```diff
@@ -38,4 +41,76 @@ public function testCanWithMultipleTransition()
...
+    }
+
+    public function testBuildTransitionBlockerListWithMultipleTransition()
+    {
+        $definition = $this->createComplexStateMachineDefinition();
+
+        $net = new StateMachine($definition);
+        $subject = new \stdClass();
+
+        // If you are in place "b" you should be able to apply "t1" and "t2"
```
Comment (stof) [inline src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: comment is not consistent with the following code

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php [hunk 2]
```diff
@@ -38,4 +41,76 @@ public function testCanWithMultipleTransition()
...
+
+        $subject = new \stdClass();
+
+        // When there are multiple transitions with the same name then method buildTransitionBlockerList might
+        // return result of different transition than expected.
+        // For example: buildTransitionBlockerList foreach two transitions "t1" where one starts in place "a" and second in place "d" and
+        // we are currently in place "a". Method buildTransitionBlockerList first processes transition "t1" with from place "a"
+        // but there is no break so it continues to transition "t1" with from place "d" where it exits loop because there are no more
+        // transitions with name "t1". So returned transition blocker list contains result of transition "t1" with from place "d" because
+        // it was executed lastest. This result is incorrect because we got "Blocked by marking" transition blocker of transition "t1"
```
Comment (stof) [inline src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: typo `latest`

File: src/Symfony/Component/Workflow/Workflow.php [hunk 3]
```diff
@@ -118,8 +118,15 @@ public function buildTransitionBlockerList($subject, string $transitionName): Tr
...
-            if ($transitionBlockerList->isEmpty()) {
-                continue;
+            // For case of non unique transition names exit the loop on transition that is supposed to be validated.
+            // Else we might get different blocker list containing blockedByMarking instead of blocker from the guard event.
+            // Reason is if you have transition "t" from "a" to "c" and transition "t" from "b" to "c" and you are in place "a"
+            // transitions loop might exit in transition with "from" place "b"
+            // which means you would get $transitionBlockerList containing blockedByMarking instead of guard blocker or none.
+            foreach ($transition->getFroms() as $place) {
+                if ($marking->has($place)) {
+                    break 2;
```
Comment (stof) [inline src/Symfony/Component/Workflow/Workflow.php]: What happens if the transition for the current place comes *after* the one one in the list of transitions ?

I think this marking check should happen *before* building the blocker list, and use `continue` rather than `break`

**Comment by Tetragramat** [comment]
@xabbuh's solution looks better than my hotfix.

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php (hunk 1 above)
Comment (Tetragramat) [reply src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: I did not notice that while copying and modifying testCanWithMultipleTransition. There is the same inconsistency.

**Comment by xabbuh** [comment]
@Tetragramat feel free to just copy the changes into your PR

File: src/Symfony/Component/Workflow/Workflow.php (hunk 3 above)
Comment (Tetragramat) [reply src/Symfony/Component/Workflow/Workflow.php]: There is break instead of continue because if you are making completely invalid transition (transition that exists, but you are not in valid state to execute that transition) then you would not get blocked by marking transition blocker but UndefinedTransitionException.

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php [hunk 4]
```diff
@@ -38,4 +41,76 @@ public function testCanWithMultipleTransition()
...
+        $subject->marking = 'a';
+        $this->assertTrue($net->buildTransitionBlockerList($subject, 't1')->isEmpty());
+        $subject->marking = 'd';
+        $this->assertTrue($net->buildTransitionBlockerList($subject, 't1')->isEmpty());
+
+        $subject->marking = 'b';
+        $this->assertFalse($net->buildTransitionBlockerList($subject, 't1')->isEmpty());
+    }
+
+    public function testBuildTransitionBlockerListWithMultipleTransition()
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: `testBuildTransitionBlockerListWithMultipleTransitions`

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php [hunk 5]
```diff
@@ -38,4 +41,76 @@ public function testCanWithMultipleTransition()
...
     }
+
+    public function testBuildTransitionBlockerList()
+    {
+        $definition = $this->createComplexStateMachineDefinition();
+
+        $net = new StateMachine($definition);
+        $subject = new \stdClass();
+
+        // If you are in place "a" you should be able to apply "t1"
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: not sure this adds that much value, also it may be confusing why the comment is only here but below for `d` and `b`

File: src/Symfony/Component/Workflow/Tests/WorkflowTest.php [hunk 6]
```diff
@@ -195,6 +195,34 @@ public function testBuildTransitionBlockerListReturnsUndefinedTransition()
...
+        $workflow = new Workflow($definition, new MultipleStateMarkingStore());
+
+        $this->assertTrue($workflow->buildTransitionBlockerList($subject, 't1')->isEmpty());
+        $this->assertFalse($workflow->buildTransitionBlockerList($subject, 't2')->isEmpty());
+
+        $subject->marking = array('b' => 1);
+
+        $this->assertFalse($workflow->buildTransitionBlockerList($subject, 't1')->isEmpty());
+        // In a workflow net, all "from" places should contain a token to enable
+        // the transition.
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/Tests/WorkflowTest.php]: How is this comment related to the code being executed (maybe I just misunderstand what this tries to explain)?

File: src/Symfony/Component/Workflow/Tests/StateMachineTest.php [hunk 7]
```diff
@@ -38,4 +41,76 @@ public function testCanWithMultipleTransition()
...
+
+        // When there are multiple transitions with the same name then method buildTransitionBlockerList might
+        // return result of different transition than expected.
+        // For example: buildTransitionBlockerList foreach two transitions "t1" where one starts in place "a" and second in place "d" and
+        // we are currently in place "a". Method buildTransitionBlockerList first processes transition "t1" with from place "a"
+        // but there is no break so it continues to transition "t1" with from place "d" where it exits loop because there are no more
+        // transitions with name "t1". So returned transition blocker list contains result of transition "t1" with from place "d" because
+        // it was executed latest. This result is incorrect because we got "Blocked by marking" transition blocker of transition "t1"
+        // with from place "d" instead of guard blocker "Transition blocker of place a" of transition "t1" with from place "a".
+        // This test checks if this bug does not happen.
```
Comment (xabbuh) [inline src/Symfony/Component/Workflow/Tests/StateMachineTest.php]: while this is correct, I would remove the comment above and only add a comment below that reads like this:

```php
// There may be multiple transitions with the same name. Make sure that transitions
// that are not enabled by the marking are evaluated.
// see https://github.com/symfony/symfony/issues/28432
```

**Review by xabbuh — APPROVED** [review]
with some minor comments

**Review by Nyholm — APPROVED** [review]
I’m happy with this PR after the suggested fixes.

File: src/Symfony/Component/Workflow/Tests/WorkflowTest.php (hunk 6 above)
Comment (xabbuh) [reply src/Symfony/Component/Workflow/Tests/WorkflowTest.php]: if you want to, you can update that test too

Comment (Tetragramat) [reply src/Symfony/Component/Workflow/Tests/WorkflowTest.php]: I'll let it be.

**Comment by nicolas-grekas** [comment]
friendly ping @lyrixx

**Comment by lyrixx** [comment]
Hello. Sorry for the very late reply. I will have a look tomorrow or after tomorrow. Thanks

**Comment by lyrixx** [comment]
Hello, 

Thanks for you work. It's :+1: I made update the code to make it simpler in #29141

**Comment by nicolas-grekas** [comment]
Closing in favor of #29141, thanks @Tetragramat
